src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs

```diff
@@ -0,0 +1,628 @@
+                    }
+
+                    return (resultCollection, collectionItem, index) =>
+                    {
+                        addMethod.Invoke(resultCollection, new object[] { collectionItem });
+                        return resultCollection;
+                    };
+                }
+            }
+        }
+
+        private object ReflectionReadCollectionItem(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, CollectionDataContract collectionContract, Type itemType, string itemName, string itemNs,
+            ref ReflectionReadValueDelegate reflectionReadValueDelegate)
+        {
+            if (collectionContract.Kind == CollectionKind.Dictionary || collectionContract.Kind == CollectionKind.GenericDictionary)
+            {
+                return ReflectionReadDictionaryItem(xmlReader, context, collectionContract);
+            }
+            else
+            {
+                reflectionReadValueDelegate = reflectionReadValueDelegate ?? GetReflectionReadValueDelegate(xmlReader, context, itemType, itemName, itemNs);
+                return reflectionReadValueDelegate(xmlReader, context, itemType, itemName, itemNs);
+            }
+        }
+
+        private bool ReflectionTryReadPrimitiveArray(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString collectionItemName, XmlDictionaryString collectionItemNamespace, Type type, Type itemType, int arraySize, out object resultArray)
+        {
+            resultArray = null;
+
+            PrimitiveDataContract primitiveContract = PrimitiveDataContract.GetPrimitiveDataContract(itemType);
+            if (primitiveContract == null)
+                return false;
+
+            switch (itemType.GetTypeCode())
+            {
+                case TypeCode.Boolean:
+                    {
+                        bool[] boolArray = null;
+                        xmlReader.TryReadBooleanArray(context, collectionItemName, collectionItemNamespace, arraySize, out boolArray);
+                        resultArray = boolArray;
+                    }
+                    break;
+                case TypeCode.DateTime:
+                    {
+                        DateTime[] dateTimeArray = null;
+                        var jsonReader = xmlReader as JsonReaderDelegator;
+                        if (jsonReader != null)
+                        {
+                            jsonReader.TryReadJsonDateTimeArray(context, collectionItemName, collectionItemNamespace, arraySize, out dateTimeArray);
+                        }
+                        else
+                        {
+                            xmlReader.TryReadDateTimeArray(context, collectionItemName, collectionItemNamespace, arraySize, out dateTimeArray);
+                        }
```

Comment:
Why not make XmlReaderDelegator.TryReadDateTimeArray virtual and override it in JsonReaderDelegator?